src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,372 @@
+  EXPECT_EQ(done_status, Status::OK);
+
+  EXPECT_EQ(callbackCalledCount, 11);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 11);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 11);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+TEST_F(ServiceControlClientImplQuotaTest, TestCachedQuotaWithStoredCallback) {
+  ServiceControlClientOptions options(CheckAggregationOptions(0, 500, 1000),
+                                      QuotaAggregationOptions(10, 500),
+                                      ReportAggregationOptions(0, 500));
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  int callbackCalledCount = 0;
+  options.quota_transport = [&callbackCalledCount](
+      const AllocateQuotaRequest& request, AllocateQuotaResponse* response,
+      TransportDoneFunc on_done) {
+    callbackCalledCount++;
+    on_done(Status::OK);
+  };
+
+  std::unique_ptr<ServiceControlClient> client_ =
+      CreateServiceControlClient(kServiceName, kServiceConfigId, options);
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&done_status](Status status) { done_status = status; });
+
+  for (int i = 0; i < 10; i++) {
+    client_->Quota(quota_request1_, &quota_response,
+                   [&done_status](Status status) { done_status = status; });
+  }
+
+  EXPECT_EQ(done_status, Status::OK);
+
+  EXPECT_EQ(callbackCalledCount, 1);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 11);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+TEST_F(ServiceControlClientImplQuotaTest, TestNonCachedQuotaWithCallback) {
+  ServiceControlClientOptions options(CheckAggregationOptions(0, 500, 1000),
+                                      QuotaAggregationOptions(0, 500),
+                                      ReportAggregationOptions(0, 500));
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.quota_transport = mock_quota_transport_.GetFunc();
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  int callbackCalledCount = 0;
+
+  std::unique_ptr<ServiceControlClient> client_ =
+      CreateServiceControlClient(kServiceName, kServiceConfigId, options);
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&done_status](Status status) { done_status = status; },
+                 [&callbackCalledCount](const AllocateQuotaRequest& request,
+                                        AllocateQuotaResponse* response,
```

Comment:
this test only verified on_done() function is called.   This is not very useful.